rmw_zenoh_cpp/src/rmw_client.cpp

```diff
@@ -425,60 +574,60 @@ rmw_take_response(
   }
-  RCUTILS_LOG_DEBUG_NAMED("rmw_zenoh_cpp", "[rmw_take_response] Response found: %s", key.c_str());
 
-  auto response_bytes = client_data->zn_response_messages_[key];
+  // NOTE(CH3): Potential place to handle "QoS" (e.g. could pop from back so it is LIFO)
+  auto response_bytes_ptr = client_data->zn_response_message_queue_.back();
+  client_data->zn_response_message_queue_.pop_back();
+
+  client_data->response_queue_mutex_.unlock();
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    "rmw_zenoh_cpp",
+    "[rmw_take] Response found: %s",
+    client_data->zn_response_topic_key_);
 
   // RETRIEVE METADATA =========================================================
   // TODO(CH3): Again, refactor this into a modular set of functions eventually
-  size_t meta_length = sizeof(rmw_client_data_t::sequence_id);
+  size_t meta_length = sizeof(std::int64_t); // Internal type of the atomic sequence ID
 
   // Use metadata
-  memcpy(&request_header->request_id.sequence_number,
-         &response_bytes.back() + 1 - meta_length,
-         meta_length);
+  memcpy(
+    &request_header->request_id.sequence_number,
+    &response_bytes_ptr->back() + 1 - meta_length,
+    meta_length);
 
   // DESERIALIZE MESSAGE =======================================================
-  size_t data_length = response_bytes.size() - meta_length;
+  size_t data_length = response_bytes_ptr->size() - meta_length;
 
   unsigned char * cdr_buffer = static_cast<unsigned char *>(
-      allocator->allocate(data_length, allocator->state));
-  if (!cdr_buffer) {
-    RMW_SET_ERROR_MSG("failed allocate response message bytes");
-    allocator->deallocate(cdr_buffer, allocator->state);
-    return RMW_RET_ERROR;
-  }
-  memcpy(cdr_buffer, &response_bytes.front(), data_length);
-
-  // Remove stored message after successful retrieval
-  client_data->zn_response_messages_.erase(key);
+    allocator->allocate(data_length, allocator->state));
+  memcpy(cdr_buffer, &response_bytes_ptr->front(), data_length);
 
   // Object that manages the raw buffer.
   eprosima::fastcdr::FastBuffer fastbuffer(reinterpret_cast<char *>(cdr_buffer), data_length);
 
-  // Object that deserializes the data.
-  eprosima::fastcdr::Cdr deser(fastbuffer,
-                               eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
-                               eprosima::fastcdr::Cdr::DDS_CDR);
-
+  // Object that deserializes the data
+  eprosima::fastcdr::Cdr deser(
+    fastbuffer,
+    eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
+    eprosima::fastcdr::Cdr::DDS_CDR);
   if (!client_data->response_type_support_->deserializeROSmessage(
-      deser, ros_response, client_data->response_type_support_impl_)) {
-    RMW_SET_ERROR_MSG("failed deserialize ROS response message");
+      deser, ros_response, client_data->response_type_support_impl_)
+  ) {
+    RMW_SET_ERROR_MSG("could not deserialize ROS response message");
     return RMW_RET_ERROR;
   }
 
   *taken = true;
-  allocator->deallocate(cdr_buffer, allocator->state);
```

Comment:
Where is this getting deleted now?

Reply:
Not sure how this happened.

Fixed, here and everywhere. (e7a84d88577d0b063c99dad36a8136574973e2cc)